src/nix-collect-garbage/nix-collect-garbage.cc

```diff
@@ -0,0 +1,89 @@
+#include "hash.hh"
+#include "shared.hh"
+#include "globals.hh"
+
+#include <iostream>
+
+using namespace nix;
+
+const std::string profilesDir = "/nix/var/nix/profiles";
+
+std::string gen = "";
+bool dryRun = false;
+
+void runProgramSimple(Path program, const Strings & args)
+{
+    checkInterrupt();
+
+    /* Fork. */
+    Pid pid = startProcess([&]() {
+        Strings args_(args);
+        args_.push_front(program);
+        auto cargs = stringsToCharPtrs(args_);
+
+        execv(program.c_str(), (char * *) &cargs[0]);
+
+        throw SysError(format("executing ‘%1%’") % program);
+    });
+
+    pid.wait(true);
+}
+
+
+/* If `-d' was specified, remove all old generations of all profiles.
+ * Of course, this makes rollbacks to before this point in time
+ * impossible. */
+
+void removeOldGenerations(std::string dir)
+{
+    for (auto & i : readDirectory(dir)) {
+        checkInterrupt();
+
+        auto path = dir + "/" + i.name; 
+        auto type = getFileType(path);
+
+        if (type == DT_LNK) {
+            auto link = readLink(path);
+            if (link.find("link") != string::npos) {
+                printMsg(lvlInfo, format("removing old generations of profile %1%") % path);
+
+                runProgramSimple(settings.nixBinDir + "/nix-env", Strings{"-p", path, "--delete-generations", gen, dryRun ? "--dry-run" : ""});
```

Comment:
Rather than calling `nix-env`, it would be cleaner to factor out opDeleteGenerations() from nix-env.cc into a library function and call it directly. IMHO that's one of the main benefits of rewriting in C++ (being able to call APIs directly rather than going through command line invocations).